pppd/sys-linux.c

```diff
@@ -300,108 +300,143 @@ extern int dfl_route_metric;
-     * So ignore return value from the following setsockopt() call as setting
-     * option NETLINK_CAP_ACK means for us just a kernel hint / optimization.
-     */
-    one = 1;
-    setsockopt(fd, SOL_NETLINK, NETLINK_CAP_ACK, &one, sizeof(one));
+        /*
+         * Tell kernel to not send to us payload of acknowledgment error message.
+         * NETLINK_CAP_ACK option is supported since Linux kernel version 4.3 and
+         * older kernel versions always send full payload in acknowledgment netlink
+         * message. We ignore payload of this message as we need only error code,
+         * to check if our set remote peer address request succeeded or failed.
+         * So ignore return value from the following setsockopt() call as setting
+         * option NETLINK_CAP_ACK means for us just a kernel hint / optimization.
+         */
+        one = 1;
+        setsockopt(fd, SOL_NETLINK, NETLINK_CAP_ACK, &one, sizeof(one));
 
-    memset(&nladdr, 0, sizeof(nladdr));
-    nladdr.nl_family = AF_NETLINK;
+        memset(&nladdr, 0, sizeof(nladdr));
+        nladdr.nl_family = AF_NETLINK;
 
-    if (bind(fd, (struct sockaddr *)&nladdr, sizeof(nladdr)) < 0) {
-        error("send_rtnetlink_msg: bind(AF_NETLINK): %m (line %d)", __LINE__);
-        close(fd);
-        return 1;
+        if (bind(fd, (struct sockaddr *)&nladdr, sizeof(nladdr)) < 0) {
+            error("rtnetlink_msg: bind(AF_NETLINK): %m (line %d)", __LINE__);
+            close(fd);
+            return 1;
+        }
+
+        if (shared_fd)
+            *shared_fd = fd;
     }
 
     memset(&nladdr, 0, sizeof(nladdr));
     nladdr.nl_family = AF_NETLINK;
 
-    memset(&iov, 0, sizeof(iov));
-    iov.iov_base = nlreq;
-    iov.iov_len = nlreq_len;
+    memset(&iov[0], 0, sizeof(iov[0]));
+    iov[0].iov_base = nlreq;
+    iov[0].iov_len = nlreq_len;
 
     memset(&msg, 0, sizeof(msg));
     msg.msg_name = &nladdr;
     msg.msg_namelen = sizeof(nladdr);
-    msg.msg_iov = &iov;
+    msg.msg_iov = &iov[0];
     msg.msg_iovlen = 1;
 
     if (sendmsg(fd, &msg, 0) < 0) {
-        error("send_rtnetlink_msg: sendmsg(%s): %m (line %d)", desc, __LINE__);
-        close(fd);
+        error("rtnetlink_msg: sendmsg(%s): %m (line %d)", desc, __LINE__);
+        if (!shared_fd)
+            close(fd);
         return 1;
     }
 
-    memset(&iov, 0, sizeof(iov));
-    iov.iov_base = &nlresp;
-    iov.iov_len = sizeof(nlresp);
+    memset(iov, 0, sizeof(iov));
+    iov[0].iov_base = &nlresp_hdr;
+    if (nlresp_size && *nlresp_size > sizeof(nlresp_hdr)) {
+        iov[0].iov_len = offsetof(struct nlresp_hdr, nlerr);
+        iov[1].iov_base = nlresp_data;
+        iov[1].iov_len = *nlresp_size;
+    } else {
+        iov[0].iov_len = sizeof(nlresp_hdr);
+    }
 
     memset(&msg, 0, sizeof(msg));
     msg.msg_name = &nladdr;
     msg.msg_namelen = sizeof(nladdr);
-    msg.msg_iov = &iov;
-    msg.msg_iovlen = 1;
+    msg.msg_iov = iov;
```

Comment:
For example here, in other places you write &iov[0] ... it's the same thing from a pointer arithmetic perspective, but I suggest it be used consistently for the sake of readability, and &iov[0] is probably more clear about the intent (we want the address of element 0).

Reply:
If `iov` is array and you want to reference first member of this array then it is better to write `&iov[0]` to explicitly express array context. If `iov` is just ordinary pointer to structure and you want a reference to this structure then it is better to write `iov` (and not `&iov[0]`) to not confuse people about array. Of course both construction should do same thing.

And I follow these patterns. That is why on the first place I used `&iov[0]` to explicitly pass just one member (indicated by `msg.msg_iovlen = 1;`) and on second place I used `iov` to explicitly express that more members (= array; indicated via larger and non-constant value in `msg.msg_iovlen`) is passed.